nav2_controller/src/nav2_controller.cpp

```diff
@@ -305,22 +306,70 @@ void ControllerServer::setPlannerPath(const nav_msgs::msg::Path & path)
     end_pose.pose.position.x, end_pose.pose.position.y);
 }
 
+double getDistance(
+  const geometry_msgs::msg::PoseStamped & pose_a,
+  const geometry_msgs::msg::PoseStamped & pose_b)
+{
+  double x_diff = fabs(pose_a.pose.position.x - pose_b.pose.position.x);
+  double y_diff = fabs(pose_a.pose.position.y - pose_b.pose.position.y);
+
+  return std::hypot(x_diff, y_diff);
+}
+
+geometry_msgs::msg::TwistStamped getTwistToPose(
+  const geometry_msgs::msg::PoseStamped & pose_from,
+  const geometry_msgs::msg::PoseStamped & pose_to)
+{
+  double x_diff = pose_to.pose.position.x - pose_from.pose.position.x;
+  double y_diff = pose_to.pose.position.y - pose_from.pose.position.y;
+
+  tf2::Quaternion q;
+  tf2::fromMsg(pose_from.pose.orientation, q);
+  tf2::Matrix3x3 m(q);
+  double roll, pitch, yaw;
+  m.getRPY(roll, pitch, yaw);
+
+  geometry_msgs::msg::TwistStamped ret;
+  ret.header = pose_from.header;
+
+  // Get the angular speed (todo: control max vel)
+  double w = atan2(y_diff, x_diff) - yaw;
+  // Normalize [-pi, pi)
+  w = atan2(sin(w), cos(w));
+
+  ret.twist.angular.z = w;
+
+  return ret;
+}
+
+
 void ControllerServer::computeAndPublishVelocity()
 {
   geometry_msgs::msg::PoseStamped pose;
+  geometry_msgs::msg::PoseStamped last_path_pose = 
+    action_server_->get_current_goal()->path.poses.back();
+  double distance_to_goal = action_server_->get_current_goal()->distance_to_goal;
+  bool keep_running = action_server_->get_current_goal()->keep_running;
 
   if (!getRobotPose(pose)) {
     throw nav2_core::PlannerException("Failed to obtain robot pose");
   }
 
-  progress_checker_->check(pose);
-
+  if (!keep_running) {
```

Comment:
I'm not also sure about this keep running thing - if you've made it to `N` meters away from your goal, you would want it to stop. Once it starts moving again, then the controller can follow the new path to be `N` meters from that again.

Reply:
I think the robot should slow down as it approaches the target so that you can keep smooth motion, else I think it would be pretty jerky